runners: stm32cubeprogrammer: add support for --dev-id

mathieuchopstm opened the pull request:
Add support for the standard `--dev-id` argument used to select target ST-Link debug probe based on serial number. Note that this could already be achieved using custom argument `--conn-modifiers`. (`--dev-id <XXX>` is equivalent to `--conn-modifiers sn=<XXX>`)

mathieuchopstm committed:
runners: stm32cubeprogrammer: add support for --dev-id

Add support for the standard `--dev-id` argument used to select
target ST-Link debug probe based on serial number. Note that this
could already be achieved using custom argument `--conn-modifiers`.
(`--dev-id <XXX>` is equivalent to `--conn-modifiers sn=<XXX>`)

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

erwango reviewed (commented):
Let's migrate [Twister hardware adapter script](https://github.com/zephyrproject-rtos/zephyr/blob/b028e7cfda74da8a71ba8b505f3ba16fe4fbc74a/scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py#L109) to make use of this new option instead of current tool-opt.

erwango reviewed (changes requested):
Making https://github.com/zephyrproject-rtos/zephyr/pull/100220#pullrequestreview-3519543530 blocking.

mathieuchopstm committed:
scripts: pytest: hw_adapter: use --dev-id for STM32CubeProgrammer runner

Now that the STM32CubeProgrammer runner supports the standard `--dev-id`
argument, replace usage of `--tool-opt` in the pytest harness hardware
adapter.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
twister: device-handler: use `--dev-id` for STM32CubeProgrammer runner

Now that the STM32CubeProgrammer runner supports the standard `--dev-id`
argument, replace usage of `--tool-opt` in the Twister device handler code.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm commented:
About CI failures:
>```
>E0611:No name 'ZEPHYR_BASE' in module 'twister' (no-name-in-module)
>File:scripts/tests/twister/test_handlers.py
>Line:35
>Column:0
>Error: Check python files with ruff.
>```

is entirely out of my control. How can I override CI to ignore this false positive? Or is it a legitimate error?

>```
>Python lint error (SIM114) see https://docs.astral.sh/ruff/rules/if-with-same-arms:Combine `if` branches using logical `or` operator
>File:/home/runner/work/zephyr/zephyr/scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py
>Line:106
>Column:13
>EndLine:111
>EndColumn:43
>```

I understand CI wants
```diff
-          elif runner == 'jlink':
+          elif runner == 'jlink' or (runner == 'stm32cubeprogrammer' and self.device_config.product != "BOOT-SERIAL"):
                base_args.append('--dev-id')
                base_args.append(board_id)
-            elif runner == 'stm32cubeprogrammer' and self.device_config.product != "BOOT-SERIAL":
-                base_args.append('--dev-id')
-                base_args.append(board_id)
```

I have no problem with this but it would create an inconsistency. Should I do it anyways? Or should another solution be adopted? One idea that comes to mind is unifying all "conforming" runners under a single `else` using `--dev-id`:

<details>

```diff
        if board_id := self.device_config.id:
            if runner == 'pyocd':
                extra_args.append('--board-id')
                extra_args.append(board_id)
            elif runner == "esp32":
                extra_args.append("--esp-device")
                extra_args.append(board_id)
-           elif runner in ('nrfjprog', 'nrfutil', 'nrfutil_next'):
-              extra_args.append('--dev-id')
-               extra_args.append(board_id)
            elif runner == 'openocd' and self.device_config.product in ['STM32 STLink', 'STLINK-V3']:
                extra_args.append('--cmd-pre-init')
                extra_args.append(f'hla_serial {board_id}')
            elif runner == 'openocd' and self.device_config.product == 'EDBG CMSIS-DAP':
                extra_args.append('--cmd-pre-init')
                extra_args.append(f'cmsis_dap_serial {board_id}')
            elif runner == "openocd" and self.device_config.product == "LPC-LINK2 CMSIS-DAP":
                extra_args.append("--cmd-pre-init")
                extra_args.append(f'adapter serial {board_id}')
-           elif runner == 'jlink':
-               base_args.append('--dev-id')
-               base_args.append(board_id)
            elif runner == 'stm32cubeprogrammer' and self.device_config.product != "BOOT-SERIAL":
                base_args.append('--dev-id')
                base_args.append(board_id)
            elif runner == 'linkserver':
                base_args.append(f'--probe={board_id}')
+           else:
+               # 'nrfjprog', 'nrfutil', 'nrfutil_next', 'jlink'
+               base_args.append('--dev-id')
+               base_args.append(board_id)
```

(We could also hardcode `BOOT-SERIAL` as a special case in the `stm32cubeprogrammer` runner to avoid needing a separate branch)

</details>

I'll let maintainers decide :slightly_smiling_face:

marc-hb commented:
> is entirely out of my control. How can I override CI to ignore this false positive? Or is it a legitimate error? [...] I understand CI wants [...]

Unlike some other projects, Zephyr CI is not a distinct entity with opaque magic and other secrets. Everything is transparent and reviewed as the rest of the code and test code. That's probably the main reason why Zephyr CI has been so successful. For instance this one:

```
/scripts/ci/check_compliance.py -m pylint -c HEAD~3..HEAD -v DEBUG
INFO    : pylint --output-format=json2 --rcfile=...

1 check(s) failed
ERROR   : Test Pylint failed: 
E0611:No name 'ZEPHYR_BASE' in module 'twister' (no-name-in-module)
File:scripts/tests/twister/test_handlers.py
Line:35
```

I agree it seems unrelated to your PR. I'm guessing the reason this wasn't noticed before is because:
1. This file has not changed recently (only changed filed are scanned by pylint)
2. someone or something upgraded pylint or changed pylintrc without re-scanning all the files.

I don't know about this one specifically but `git grep` finds a good number of `# pylint: disable=no-name-in-module`, so this would not be the first false positive. Maybe you can submit that `disable` in a separate PR and discuss there?

Unfortunately, changing some files often requires cleaning them up a bit first. This is a community project.

mathieuchopstm commented:
> > is entirely out of my control. How can I override CI to ignore this false positive? Or is it a legitimate error? [...] I understand CI wants [...]
> 
> Unlike some other projects, Zephyr CI is not a distinct entity with opaque magic and other secrets. Everything is transparent and reviewed as the rest of the code and test code. That's probably the main reason why Zephyr CI has been so successful. For instance this one:

Right, I may have been overdramatic :sweat_smile:
What I meant to say is that the CI failure is not directly caused by my change (i.e., regardless of what I change, the warning will be triggered)

> I don't know about this one specifically but `git grep` finds a good number of `# pylint: disable=no-name-in-module`, so this would not be the first false positive. Maybe you can submit that `disable` in a separate PR and discuss there?

Thanks for the hint - I'll try adding it as a preliminary commit.

mathieuchopstm committed:
twister: device-handler: silence no-name-in-module warning

Silence CI warning caused by line "from . import ZEPHYR_BASE". This is a
legitimate pattern, but pylint seems unable to deduce so. Many other
scripts perform a similar silencing.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>
